src/librbd/group/ListSnapshotsRequest.h

```diff
@@ -0,0 +1,97 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_GROUP_LIST_SNAPSHOTS_REQUEST_H
+#define CEPH_LIBRBD_GROUP_LIST_SNAPSHOTS_REQUEST_H
+
+#include "include/int_types.h"
+#include "include/types.h"
+#include "include/rados/librados.hpp"
+#include "cls/rbd/cls_rbd_types.h"
+
+#include <string>
+#include <vector>
+
+class Context;
+
+namespace librbd {
+
+struct ImageCtx;
+
+namespace group {
+
+template <typename ImageCtxT = librbd::ImageCtx>
+class ListSnapshotsRequest {
+public:
+  static ListSnapshotsRequest *create(
+      librados::IoCtx &group_io_ctx, const std::string &group_id, bool sort,
+      bool require_sort, std::vector<cls::rbd::GroupSnapshot> *snaps,
+      Context *on_finish) {
+    return new ListSnapshotsRequest(group_io_ctx, group_id, sort, require_sort,
+                                    snaps, on_finish);
+  }
+
+  ListSnapshotsRequest(librados::IoCtx &group_io_ctx,
+                       const std::string &group_id,
+                       bool sort, bool require_sort,
+                       std::vector<cls::rbd::GroupSnapshot> *snaps,
+                       Context *on_finish)
+    : m_group_io_ctx(group_io_ctx), m_group_id(group_id), m_sort(sort),
+      m_require_sort(require_sort), m_snaps(snaps), m_on_finish(on_finish) {
+  }
+
+  void send();
+
+private:
+  /**
+   * @verbatim
+   *
+   * <start>    /--------\
+   *    |       |        | (if required. repeat if more
+   *    v       v        |  entries)
+   *  LIST_SNAP_ORDERS --/
+   *    |       /--------\
+   *    |       |        | (repeat if more
+   *    v       v        |  snapshots)
+   *  LIST_SNAPSHOTS ----/
+   *    |
+   *    v
+   *  SORT_SNAPS (if required)  
+   *    |
+   *    v
+   *  <finish>
+   *
+   * @endverbatim
+   */
+
+  librados::IoCtx &m_group_io_ctx;
+  const std::string m_group_id;
+  bool m_sort;
+  bool m_require_sort; //  Fail if true and unable to sort
+  std::vector<cls::rbd::GroupSnapshot> *m_snaps;
+  std::map<std::string, uint64_t> m_snap_orders;
+  Context *m_on_finish;
+  bool m_fail_if_no_sort = false;
+
+  cls::rbd::GroupSnapshot m_start_after;
+  std::string m_start_after_order;
```

Comment:
These fields seem redundant to me -- `list_snap_orders()` could simply take a string (i.e. have either `""` or `snap_orders.rbegin()->first` passed to it) and similar for `list_snapshots()`.  It would be more explicit and spare a bit of memory too.

Reply:
I prefer to keep this as it is to stay consistent with similar implementations across the code.

Reply:
Can you point me at the implementation(s) that you are looking at?  Specifically, an implementation where there is a similar loop over a single state, not across multiple states.

Reply:
`librbd::image::GetMetadataRequest<I>::metadata_list()`
`rbd::mirror::pool_watcher::RefreshImagesRequest<I>::mirror_image_list()`
`rbd::mirror::image_map::LoadRequest<I>::mirror_image_list()

The next key is always a class member, not an argument to the method. This looks much cleaner IMO.

Reply:
> `librbd::image::GetMetadataRequest<I>::metadata_list()` 

This is because `last_key` is an argument to the whole state machine.

`rbd::mirror::pool_watcher::RefreshImagesRequest<I>::mirror_image_list()` 

This one appears to be fine, but it's something I'd fix as a cleanup.

`rbd::mirror::image_map::LoadRequest<I>::mirror_image_list()`

This one actually seems to have a bug.  In case there are more than `MAX_RETURN` images, `mirror_image_list()` starts where `image_map_list()` previously started, not from scratch.  I doubt it's the intended behavior...

> 
> The next key is always a class member, not an argument to the method. This looks much cleaner IMO.

I disagree.  A class member is a bit like a global variable -- one must keep the entire class in mind to know which values are possible at which point, remember to reset it when needed, there might be concurrency issues, etc.  If one encounters a place where it's not reset (like in `LoadRequest::mirror_image_list()`), they have to go through the entire class again, trying to guess whether it's the intended behavior or not.

Overall, not only it's not cleaner, it actually increases the potential for bugs.  Any time a local variable or passing a parameter is sufficient, that is what should be used.

Reply:
> This one actually seems to have a related bug. In case there are more than `MAX_RETURN` images, `mirror_image_list()` starts where `image_map_list()` previously started, not from scratch. I doubt it's the intended behavior...

Filed https://tracker.ceph.com/issues/67650

Reply:
I disagree with this - currently refreshing the changes as is.

Reply:
> I disagree with this - currently refreshing the changes as is. We can discuss this further.

I presented a technical argument, backed by an instance of a bug of _exactly_ the "something that doesn't need to be a class member was made a class member anyway" nature discovered on the spot in one of the two existing implementations that you pointed out.  (I'm saying two instead of three because the use of the class member in `GetMetadataRequest` can be justified at least to some (small) degree.)  The bug went unnoticed for several releases most likely because hitting it requires more than 1024 images being mirrored which we don't test, but it would have been avoided _by construction_ had a parameter been used.

Nothing other than "this looks much cleaner" was presented in favor of the code as is.  If you feel like there is more discussion to be had here, that would be the place to start.